internal/controllers/addon_instance_controller.go

```diff
@@ -0,0 +1,71 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+	addoninstanceapi "github.com/openshift/addon-operator/pkg/addoninstance"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type AddonInstanceReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+
+	HeartbeatCheckerRate time.Duration
+}
+
+func (r *AddonInstanceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&addonsv1alpha1.AddonInstance{}).
+		Complete(r)
+}
+
+// AddonInstanceReconciler/Controller entrypoint
+func (r *AddonInstanceReconciler) Reconcile(
+	ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("addoninstance", req.NamespacedName.String())
+
+	addonInstance := &addonsv1alpha1.AddonInstance{}
+	// r.Get() is backed by a Kubernetes Client backed by cached reads. Ref: https://github.com/kubernetes-sigs/controller-runtime/blob/v0.10.2/pkg/cluster/cluster.go#L51-L55
+	if err := r.Get(ctx, req.NamespacedName, addonInstance); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	now, lastHeartbeatTime := metav1.Now(), addonInstance.Status.LastHeartbeatTime
+	if lastHeartbeatTime.IsZero() {
+		// is it worth logging this, considering the immense amounts of logs which are going to end up getting generated?
+		//log.Info(fmt.Sprintf("skipping heartbeat for %s/%s because no lastHeartbeatTime was found.", addonInstance.Namespace, addonInstance.Name))
```

Comment:
Nope, it's a perfect valid case that an Addon did not yet implement Status reporting, so the AddonInstance Status would be empty. No need to log this every time.